lib/chef-acceptance/chef_runner.rb

```diff
@@ -3,80 +3,48 @@
 require 'json'
 require 'bundler'
 require 'chef-acceptance/acceptance_cookbook'
+require 'chef-acceptance/shellout_builder'
 
 module ChefAcceptance
   class ChefRunner
     attr_reader :acceptance_cookbook
-
     attr_reader :test_suite
+    attr_reader :shellout_builder
 
-    attr_accessor :run_recipes
-
-    def initialize(test_suite, options = {})
+    def initialize(test_suite)
       @test_suite = test_suite
       @acceptance_cookbook = test_suite.acceptance_cookbook
-      @run_recipes = options.fetch(:run_recipes, [])
-    end
-
-    def run
-      verify_run_prerequisites
-      create_required_files
-      execute_run_command
+      @shellout_builder = ShelloutBuilder.new(cwd: acceptance_cookbook.root_dir)
     end
 
-    def verify_run_prerequisites
+    def run(recipe)
       unless test_suite.exist?
-        fail <<-EOS
-Could not find test suite '#{test_suite.name}' in the current working \
-directory '#{Dir.pwd}'.
-EOS
+        fail <<-EOS.gsub(/^\s+/, "")
+          Could not find test suite '#{test_suite.name}' in the current working directory '#{Dir.pwd}'.
+        EOS
       end
 
       unless ExecutableHelper.executable_installed? 'chef-client'
         fail "Could not find chef-client in #{ENV['PATH']}"
       end
-    end
 
-    def create_required_files
-      FileUtils.rmtree temp_dir
-      FileUtils.mkpath temp_dir
-      File.write(dna_json_file, JSON.pretty_generate(dna))
+      # prep and create chef attribute and config file
+      prepare_required_files
 
-      FileUtils.mkpath chef_dir
-      File.write(chef_config_file, chef_config_template)
-    end
+      shellout_builder.with_chef_config_file(chef_config_file)
+      shellout_builder.with_dna_json_file(dna_json_file)
+      shellout_builder.with_recipe("#{AcceptanceCookbook::ACCEPTANCE_COOKBOOK_NAME}::#{recipe}")
+      chef_shellout = shellout_builder.build
 
-    def execute_run_command
-      Bundler.with_clean_env do
-        chef_zero = Mixlib::ShellOut.new(construct_run_command,
-                                        cwd: acceptance_cookbook.root_dir,
-                                        live_stream: $stdout)
+      stats = { duration: 0, error: nil }
 
-        chef_zero.run_command
-
-        chef_zero.error!
-      end
-    end
-
-    def construct_run_command
-      shellout = []
-      shellout << 'chef-client -z'
-      shellout << "-c #{File.expand_path(chef_config_file)}"
-      shellout << '--force-formatter'
-
-      if File.exist? dna_json_file
-        shellout << "-j #{File.expand_path(dna_json_file)}"
-      end
-
-      unless run_recipes.empty?
-        run_list = run_recipes.collect do |recipe|
-          "#{AcceptanceCookbook::ACCEPTANCE_COOKBOOK_NAME}::#{recipe}"
-        end
-
-        shellout << "-o #{run_list.join(',')}"
+      Bundler.with_clean_env do
+        chef_shellout.run_command
+        stats[:duration] = chef_shellout.execution_time
+        stats[:error] = chef_shellout.stdout if chef_shellout.error?
```

Comment:
I think capturing full `:stdout` output as the error will be too long and some things might go into `:stderr` anyways. 

IMO we should standardize on how we are communicating errors out of this class. I suggest we use Exceptions because we have two types of error cases that should be handled in different places:

1. Errors that will cause abortion of acceptance command: e.g. chef-client does not exist
2. Errors that will cause abortion of the suite run: e.g. test itself failed